### PR TITLE
Add MySQL protocol layer code

### LINKED ISSUE #88: Add  MySQL protocol layer code
In TinySQL, we hope everyone can focus on the design and implementation of the SQL layer. The content of the MySQL protocol layer is complex and messy, and it is not our focus. So we will provide a simple implementation at the beginning.

### PR DESCRIPTION
**Pull request by rebelice**
This closes #88 

This PR contains the simple MySQL protocol layer code. These codes refer to the implementation of TiDB, but are simplified and only include the necessary implementation.
It should be noted that the current implementation only supports the "mysql_native_password" auth plugin.

#### compile and run
```
make
./bin/tinysql-server
```
#### use MySQL-client
```
mysql --host 127.0.0.1 --port 4000 -u root
```

### TODO

- [ ] https://github.com/tidb-incubator/tinysql/issues/94
- [ ] https://github.com/tidb-incubator/tinysql/issues/96
- [ ] https://github.com/tidb-incubator/tinysql/issues/95

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+*.dylib
+
+# Test binary, built with `go test -c`
+*.test
+
+# Output of the go coverage tool, specifically when used with LiteIDE
+*.out
+
+# Dependency directories (remove the comment below to include it)
+# vendor/
```
Comment (tisonkun) [inline .gitignore]: I don't see any of these lines necessary.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include Makefile.common
+
+.PHONY: all clean test gotest server dev benchkv benchraw check checklist parser tidy ddltest
```
Comment (tisonkun) [inline Makefile]: Please remove unnecessary items at the moment.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+default: server buildsucc
+
+PROJECT=tinysql
+GOPATH ?= $(shell go env GOPATH)
+P=8
+
+# Ensure GOPATH is set before running build process.
+ifeq "$(GOPATH)" ""
+  $(error Please set the environment variable GOPATH before running `make`)
+endif
```
Comment (tisonkun) [inline Makefile]: `GOPATH ?= $(shell go env GOPATH)` already ensures this?

File: Makefile.common [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+PROJECT=tinysql
+GOPATH ?= $(shell go env GOPATH)
+P=8
+
+# Ensure GOPATH is set before running build process.
+ifeq "$(GOPATH)" ""
+  $(error Please set the environment variable GOPATH before running `make`)
+endif
+FAIL_ON_STDOUT := awk '{ print } END { if (NR > 0) { exit 1 } }'
```
Comment (tisonkun) [inline Makefile.common:24]: Maybe we should remove repeats in `Makefile`?

File: Makefile.common [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+path_to_add := $(addsuffix /bin,$(subst :,/bin:,$(GOPATH))):$(PWD)/tools/bin
+export PATH := $(path_to_add):$(PATH)
+
+GO              := GO111MODULE=on go
+GOBUILD         := $(GO) build $(BUILD_FLAG) -tags codes
+GOBUILDCOVERAGE := GOPATH=$(GOPATH) cd tidb-server; $(GO) test -coverpkg="../..." -c .
+GOTEST          := $(GO) test -p $(P)
+OVERALLS        := GO111MODULE=on overalls
+STATICCHECK     := GO111MODULE=on staticcheck
+TIDB_EDITION    ?= Community
```
Comment (tisonkun) [inline Makefile.common]: Or please tidy `Makefile.common` to what we exactly want. At least I don't think we should keep this TiDB related stuff especially about edition.

File: config/config.go [hunk 6]
```diff
@@ -0,0 +1,1053 @@
...
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"github.com/pingcap/errors"
+	zaplog "github.com/pingcap/log"
+	log "github.com/sirupsen/logrus"
+	"tinysql/util/logutil"
+	//"github.com/pingcap/tidb/util/versioninfo"
```
Comment (tisonkun) [inline config/config.go:34]:
```suggestion
```

**Review by tisonkun — COMMENTED** [review]
Generally looks good. Just remind that we're developing a course instead of yet another TiDB, so please narrow the code shape as much as possible.

**Comment by rebelice** [comment]
@tisonkun Thanks for your review. I totally agree with you. At present, the MySQL protocol layer code is still too large. There are two main reasons:
1. There are some modules that have not been simplified, such as `config` and `errno`
2. The design of the `server` module is the same as TiDB, it should be simplified and refactored
Of the two, reason 1 accounts for the main part.

The reason for doing this at present is that these two parts are not actually the focus of the TinySQL course. And these two jobs are not related to subsequent courses. So I took them down as [issue](https://github.com/tidb-incubator/tinysql/issues/96) and hoped to finish it later.

Developing and submitting labs have a higher priority, I hope we can focus on them first.

**Review by tisonkun — COMMENTED** [review]
Almost looks good. Please resolve outstanding comments.

File: Makefile [hunk 7]
```diff
@@ -14,20 +14,10 @@
 
 include Makefile.common
 
-.PHONY: all clean test gotest server dev benchkv benchraw check checklist parser tidy ddltest
+.PHONY: server
 
 default: server buildsucc
 
-PROJECT=tinysql
-GOPATH ?= $(shell go env GOPATH)
```
Comment (tisonkun) [inline Makefile]: You may still keep this line.

Comment (tisonkun) [reply Makefile]: But well. It's OK to merge. If someone wants to tolerate it, file an issue.

**Pull request merged**